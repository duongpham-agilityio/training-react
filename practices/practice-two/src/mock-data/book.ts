import { Book } from 'types';

const title = `HTML/CSS book name will displayed here, test with a
long text, will display full`;

const description = `Will show full description here, with a long text, will limit about 200  character.
Then will display the show more/show less for it. You   with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You can see this ex more/show less for it. You can see this ex more/show less for it. You can see this ex You can see this ex You can see this ex more/show less for it. You can see this ex`;

const author = 'Cu.Nguyen';
const imageURL =
  'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api';

export const book: Book = {
  name: title,
  author,
  description,
  imageURL,
  id: 0,
  publishDate: new Date().getTime(),
  createdAt: new Date().getTime(),
  updatedAt: new Date().getTime(),
  deletedAt: null,
};
