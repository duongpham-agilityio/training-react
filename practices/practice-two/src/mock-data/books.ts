// Types
import { Book } from 'types';

export const books: Book[] = [
  {
    name: 'HTML/CSS Ebook 1',
    description: 'Description of some book will displayed here',
    author: 'Duong.Pham',
    imageURL:
      'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
    createdAt: 1680158351376,
    deletedAt: null,
    updatedAt: 1680158351376,
    publishDate: 1680158351376,
    id: 1,
  },
  {
    name: 'HTML/CSS Ebook dd ',
    description:
      'Description of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed heresdssdsdsdsdDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed here',
    author: 'Duong.Pham',
    imageURL:
      'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
    createdAt: 1680158351376,
    deletedAt: null,
    updatedAt: 1680603476559,
    publishDate: 1680158351376,
    id: 3,
  },
  {
    name: 'HTML/CSS Ebook 4',
    description:
      'Description of some book will displayed here ds s ssss   sss    ',
    author: 'Duong.Pham',
    imageURL:
      'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
    createdAt: 1680158351376,
    deletedAt: null,
    updatedAt: 1680571677989,
    publishDate: 1680158351376,
    id: 4,
  },
  {
    name: 'HTML/CSS Ebook 6',
    description: 'Description of some book will displayed here',
    author: 'Duong.Pham',
    imageURL:
      'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
    createdAt: 1680158351376,
    deletedAt: null,
    updatedAt: 1680158351376,
    publishDate: 1680158351376,
    id: 7,
  },
  {
    author: 'Duong Pham',
    description:
      'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
    imageURL:
      'http://books.google.com/books/content?id=Mz8EAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'LIFE',
    createdAt: 1680603732584,
    deletedAt: null,
    updatedAt: 1680603732584,
    publishDate: 1680158351376,
    id: 10,
  },
  {
    author: 'Duong Pham',
    description:
      'Established in 1911, The Rotarian is the official magazine of Rotary International and is circulated worldwide. Each issue contains feature articles, columns, and departments about, or of interest to, Rotarians. Seventeen Nobel Prize winners and 19 Pulitzer Prize winners – from Mahatma Ghandi to Kurt Vonnegut Jr. – have written for the magazine.',
    imageURL:
      'http://books.google.com/books/content?id=ckEEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'The Rotarian',
    createdAt: 1680624355336,
    deletedAt: null,
    updatedAt: 1680624355336,
    publishDate: 1680158351376,
    id: 12,
  },
  {
    author: 'asdsad',
    description: 'asdasdasdasd',
    imageURL:
      'http://books.google.com/books/content?id=WNSUweeeSIIC&printsec=frontcover&img=1&zoom=1&source=gbs_api',
    name: 'Hello',
    createdAt: 1680658712485,
    deletedAt: null,
    updatedAt: 1680658712485,
    publishDate: 1680158351376,
    id: 13,
  },
  {
    author: 'Duong Pham',
    description:
      'Popular Mechanics inspires, instructs and influences readers to help them master the modern world. Whether it’s practical DIY home-improvement tips, gadgets and digital technology, information on the newest cars or the latest breakthroughs in science -- PM is the ultimate guide to our high-tech lifestyle.',
    imageURL:
      'http://books.google.com/books/content?id=QtQDAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'Popular Mechanics2333',
    createdAt: 1680658742125,
    deletedAt: null,
    updatedAt: 1680658742125,
    publishDate: 1680158351376,
    id: 14,
  },
  {
    author: '123',
    description:
      'Flex 4 is an open-source tool that allows developers to easily add life to web applications with dynamic user features, colorful transitions, and eye-catching animations. Flex also provides powerful data handling for industrial-strength applications. We think it should be just as much fun to learn Flex as it is to use it. And we know that fun learning gets better results. Hello! Flex 4 demonstrates how to get started without getting bogged down in technical detail or academic edge cases. In this book, User Friendly cartoon characters offer commentary and snide side comments, as the book moves quickly from Hello World into practical techniques. Each one is illustrated with a hands-on example. Along the way, readers will build a unique Flex application that mashes Yahoo Maps with Twitter to keep track of friends. Purchase of the print book comes with an offer of a free PDF, ePub, and Kindle eBook from Manning. Also available is all code from the book.',
    imageURL:
      'http://books.google.com/books/content?id=DjszEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'Hello! Flex 4',
    createdAt: 1680658880446,
    deletedAt: null,
    updatedAt: 1680658880446,
    publishDate: 1680158351376,
    id: 15,
  },
  {
    author: 'dddd',
    description:
      'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
    imageURL:
      'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'LIFE23333',
    createdAt: 1680658968253,
    deletedAt: null,
    updatedAt: 1680658968253,
    publishDate: 1680158351376,
    id: 16,
  },
  {
    author: 'dddd',
    description:
      'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
    imageURL:
      'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'LIFE23333',
    createdAt: 1680658968253,
    deletedAt: null,
    updatedAt: 1680658968253,
    publishDate: 1680158351376,
    id: 17,
  },
];
