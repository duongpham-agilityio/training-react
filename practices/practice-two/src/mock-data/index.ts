export * from 'mock-data/book';
export * from 'mock-data/books';
