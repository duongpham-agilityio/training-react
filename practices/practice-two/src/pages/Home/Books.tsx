import { MouseEvent, ReactNode, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Components
import { Heading } from 'components/commons';
import { Card } from 'components';

// Constants
import { ENDPOINT, MESSAGES } from '@constants';

// Helpers
import { convertDateTimeToTimeString } from 'helpers';

// Styles
import homeStyles from 'pages/Home/index.module.css';

// Types
import { Book } from 'types';

export interface BooksProps {
  books: Book[];
  onDeleteBook: (id: number) => void;
  children?: ReactNode;
}

const Books = ({ books, children, onDeleteBook }: BooksProps) => {
  const renderBook = useCallback(
    (book: Book) => {
      const { id, name, description, createdAt, imageURL } = book;
      const deleteHandler = (event: MouseEvent) => {
        event.preventDefault();

        onDeleteBook(id || 0);
      };

      return (
        <Card
          href={`/${ENDPOINT.BOOKS}/${id}`}
          title={name}
          description={description}
          publishedDate={convertDateTimeToTimeString(createdAt)}
          imageUrl={imageURL}
          key={id}
          onDeleteBook={deleteHandler}
        />
      );
    },
    [onDeleteBook]
  );

  return (
    <div className={homeStyles.content}>
      {books.length ? (
        <>
          <div className={homeStyles.grid}>{books.map(renderBook)}</div>
          {children}
        </>
      ) : (
        <div className={homeStyles.empty}>
          <Heading
            label={MESSAGES.EMPTY_TITLE}
            className={homeStyles.emptyTitle}
            size="xl"
          />
          <p className={homeStyles.emptyDescription}>
            {MESSAGES.EMPTY_DESCRIPTION}
          </p>
        </div>
      )}
    </div>
  );
};

const areCompare = (prev: BooksProps, next: BooksProps): boolean =>
  isEqual(prev.books, next.books) && isEqual(prev.children, next.children);

export default memo(Books, areCompare);
