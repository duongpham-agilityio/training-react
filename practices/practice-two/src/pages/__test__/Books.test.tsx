// Mocks
import 'components/__test__/mocks/image.test';

import { fireEvent, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Books, { BooksProps } from 'pages/Home/Books';

// Mock data
import { books } from 'mock-data';

// Constants
import { MESSAGES } from '@constants';

const Component = (props: BooksProps) => (
  <BrowserRouter>
    <Books {...props} />
  </BrowserRouter>
);

const setup = (props: BooksProps) => render(<Component {...props} />);

describe('Books components', () => {
  it('Match to snapshot', () => {
    const { container } = setup({
      books: books.slice(0, 5),
      onDeleteBook: jest.fn(),
    });

    expect(container).toMatchSnapshot();
  });

  it('Delete book', () => {
    const { getAllByRole } = setup({
      books: books.slice(0, 5),
      onDeleteBook: jest.fn(),
    });

    const buttons = getAllByRole('button');

    fireEvent.click(buttons[1]);
  });

  it('Empty books', () => {
    const { getByText, rerender, getAllByRole } = setup({
      books: [],
      onDeleteBook: jest.fn(),
      children: <p>Hello</p>,
    });

    expect(getByText(MESSAGES.EMPTY_DESCRIPTION)).toBeInTheDocument();

    rerender(
      <Component
        onDeleteBook={jest.fn()}
        books={[
          {
            name: 'HTML/CSS Ebook 1',
            description: 'Description of some book will displayed here',
            author: 'Duong.Pham',
            imageURL:
              'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
            createdAt: 1680158351376,
            deletedAt: null,
            updatedAt: 1680158351376,
            publishDate: 1680158351376,
          },
        ]}
      >
        <p>Hello</p>
      </Component>
    );

    const buttons = getAllByRole('button');

    fireEvent.click(buttons[1]);
  });
});
