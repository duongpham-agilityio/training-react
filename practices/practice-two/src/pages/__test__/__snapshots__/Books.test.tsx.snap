// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books components Match to snapshot 1`] = `
<div>
  <div>
    <div>
      <a
        href="/books/1"
      >
        <div>
          <h2
            class="undefined undefined "
          >
            HTML/CSS Ebook 1
          </h2>
          <p>
            Description of some book will displayed here
          </p>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/next.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
          </div>
        </div>
        <div>
          <p>
            1:39 PM
          </p>
          <img
            alt="HTML/CSS Ebook 1"
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </a>
      <a
        href="/books/3"
      >
        <div>
          <h2
            class="undefined undefined "
          >
            HTML/CSS Ebook dd 
          </h2>
          <p>
            Description of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed heresdssdsdsdsdDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed here
          </p>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/next.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
          </div>
        </div>
        <div>
          <p>
            1:39 PM
          </p>
          <img
            alt="HTML/CSS Ebook dd "
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </a>
      <a
        href="/books/4"
      >
        <div>
          <h2
            class="undefined undefined "
          >
            HTML/CSS Ebook 4
          </h2>
          <p>
            Description of some book will displayed here ds s ssss   sss    
          </p>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/next.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
          </div>
        </div>
        <div>
          <p>
            1:39 PM
          </p>
          <img
            alt="HTML/CSS Ebook 4"
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </a>
      <a
        href="/books/7"
      >
        <div>
          <h2
            class="undefined undefined "
          >
            HTML/CSS Ebook 6
          </h2>
          <p>
            Description of some book will displayed here
          </p>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/next.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
          </div>
        </div>
        <div>
          <p>
            1:39 PM
          </p>
          <img
            alt="HTML/CSS Ebook 6"
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </a>
      <a
        href="/books/10"
      >
        <div>
          <h2
            class="undefined undefined "
          >
            LIFE
          </h2>
          <p>
            LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of todayâ€™s people and events. They have free access to share, print and post images for personal use.
          </p>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/next.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
          </div>
        </div>
        <div>
          <p>
            5:22 PM
          </p>
          <img
            alt="LIFE"
            src="http://books.google.com/books/content?id=Mz8EAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
          />
        </div>
      </a>
    </div>
  </div>
</div>
`;
